### PR TITLE
compile,py: fix closure and decorator

### PR DESCRIPTION
**Pull request by wetor**
After several hours of reading the Python source code, it was found that the closure issue was only caused by the lack of assignment to `Code.Cell2arg`

The submission still requires some optimization, and I will complete it in the near future. If possible, I would also like to receive some coding suggestions

### CODE REVIEW

File: compile/compile.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 The go-python Authors.  All rights reserved.
+// Copyright 2023 The go-python Authors.  All rights reserved.
```
Comment (sbinet) [inline compile/compile.go:1]: no need to update this.

File: py/code.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 The go-python Authors.  All rights reserved.
+// Copyright 2023 The go-python Authors.  All rights reserved.
```
Comment (sbinet) [inline py/code.go:1]: ditto.

File: py/code.go [hunk 3]
```diff
@@ -167,6 +164,38 @@ func NewCode(argcount int32, kwonlyargcount int32,
...
+		lnotab)
+}
+
+// Make a new code object
+func MakeCode(argcount int32, kwonlyargcount int32,
+	nlocals int32, stacksize int32, flags int32,
+	code string, consts_ Object, names []string,
+	varnames []string, freevars []string, cellvars []string,
+	filename string, name string, firstlineno int32,
+	lnotab string) *Code {
```
Comment (sbinet) [inline py/code.go:191]:
```suggestion
func MakeCode(argcount, kwonlyargcount, nlocals, stacksize, flags int32,
	code string, consts_ Object, names, varnames, freevars, cellvars []string,
	filename, name string, firstlineno int32,
	lnotab string) *Code {
```

File: py/code.go [hunk 4]
```diff
@@ -167,6 +164,38 @@ func NewCode(argcount int32, kwonlyargcount int32,
...
+		varnames,
+		freevars,
+		cellvars,
+		filename,
+		name,
+		firstlineno,
+		lnotab)
+}
+
+// Make a new code object
```
Comment (sbinet) [inline py/code.go:185]: having both `MakeCode` and `NewCode` with the same doc and signature, this needs a meaningful documentation string, explaining what's the difference between the two.

File: vm/vm_test.go [hunk 5]
```diff
@@ -19,6 +19,11 @@ func TestVm(t *testing.T) {
 	pytest.RunTests(t, "tests")
 }
 
+func TestVm2(t *testing.T) {
+	// TODO(wetor) temp test
+	pytest.RunTests(t, "tests2")
```
Comment (sbinet) [inline vm/vm_test.go:24]: why the new `tests2` directory ?

Comment (wetor) [reply vm/vm_test.go:24]: This is a temporary debug code, and I will eventually remove it

**Comment by wetor** [comment]
PTAL, fixed an issue with compiler closures, and added test cases for decorators

**Comment by wetor** [comment]
please skip the Codecov that did not pass, as this part of the code has not been covered

File: compile/compile.go [hunk 6]
```diff
@@ -479,7 +484,8 @@ func (c *compiler) makeClosure(code *py.Code, args uint32, child *compiler, qual
 		if reftype == symtable.ScopeCell {
 			arg = c.FindId(name, c.Code.Cellvars)
 		} else { /* (reftype == FREE) */
-			arg = c.FindId(name, c.Code.Freevars)
+			// using CellAndFreeVars in closures requires skipping Cellvars
+			arg = len(c.Code.Cellvars) + c.FindId(name, c.Code.Freevars)
```
Comment (wetor) [inline compile/compile.go:488]: I couldn't find out how to use zero based subscripts to access the corresponding elements in `Code.Cellvars` and `Code.Freevars` in `Frame.Localsplus` without debugging Cpython, so I used this method to directly access `Code.CellAndFreeVars`

Comment (wetor) [reply compile/compile.go:488]: Today, I dynamically debugged CPython3.4.9, and this modification is correct. For the subscripts of `FreeVars`, the length of `CellVars` needs to be added.

By the way, the complete repair of the decorator will be submitted after this PR is merged, as some modifications to the grammar are needed

File: vm/tests/decorators.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
+# Copy from Python-3.4.9\Lib\test\test_decorators.py
```
Comment (sbinet) [inline vm/tests/decorators.py:1]: please add the license blurb. ie:

```
# Copyright 2023 The go-python Authors.  All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

# Copied from Python-3.4.9\Lib\test\test_decorators.py
```

File: vm/tests/libtest.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
+# Imitate the calling method of unittest
```
Comment (sbinet) [inline vm/tests/libtest.py:1]: please add the license blurb. ie:

```
# Copyright 2023 The go-python Authors.  All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

# Imitate ...
```

**Comment by sbinet** [comment]
(otherwise LGTM, thanks.)

File: vm/tests/decorators.py (hunk 7 above)
Comment (wetor) [reply vm/tests/decorators.py]: resolved

**Pull request merged**